Client-server interaction development beggining

Reeliks opened the pull request:
Начинается полномасштабная работа над взаимодействием клиента и сервера базы данных!
* Инфо-каналы для возможности гибко менять выходную точку сообщений, работающие на mpsc
* Клиентское подключение, имеющее ключ для взаимодействия с бд и прототип настраиваемого жизненного цикла
* Рефакторинг структуры серверного конфига, добавление новых полей
* Вынесение хэндлинга консольных команд в соответствующую структуру
* Замена blaze create на blaze init, генерирующий только %.manage.blz файл со всеми доступными конфигурационными полями (manage.blz перевоплотился в main.manage.blz, имя можно менять)
* Улучшение запуска сервера путём добавления понятных сообщений об ошибках и служебной информации о самом запуске, возможность выбирать manage file при запуске без написания .manage.blz
* Другие небольшие правки в коде после рефакторинга
Мыслю над синтаксисом языка, взглянуть можно в README.md

Reeliks committed:
Improved server raising process, deleted db/create_db.rs with 'blaze create' and made 'blaze init' instead

Reeliks committed:
Conflicts resolving after merging fs

Reeliks committed:
Info Channel, ServerInstance, Connections handling, CommandHandler

Reeliks committed:
Removing formatting test

Reeliks committed:
README.md small update

BlackSoulHub commented on .github/workflows/code.yml:
```diff
@@ -22,8 +22,8 @@ jobs:
 
-      - name: Check formatting
-        run: cargo fmt -- --check
+#      - name: Check formatting
```
Why?

cogdash replied:
По какой-то причине не проходит проверку, даже когда всё нормально. Может, я чего-то не догоняю

BlackSoulHub replied:
Try to merge #38 to this branch

cogdash replied:
Ok let's try

BlackSoulHub committed:
fmt

BlackSoulHub committed:
uncomment fmt check

Reeliks committed:
Merge pull request #38 from BlackSoulHub/main

Fix fmt

Reeliks committed:
ClientConnection -> ClientSession, small rework

Reeliks committed:
conflict resolving

BlackSoulHub commented on Cargo.toml:
```diff
@@ -6,9 +6,16 @@ edition = "2021"
+http-body-util = "0.1"
+hyper = { version = "1", features = ["full"] }
+tokio = { version = "1", features = ["full"] }
+serde = { version = "1", features = ["derive"] }
```
Do not use versions like this. Specify full version. 

Theoretically it can break build on users with older lib version

BlackSoulHub commented on src/main.rs:
```diff
@@ -1,8 +1,6 @@
-    shell::handling::handle_command_arguments()?;
-    Ok(())
+#[tokio::main]
+async fn main() {
```
I think we need move main to separate crate

As far as I know, using a combined lib and executable crate has dependency issues

cogdash replied:
I haven't heard about this approach and the problems without it. What advantages does it have?

BlackSoulHub replied:
> I haven't heard about this approach and the problems without it. What advantages does it have?

Users of this crate (as lib) also download dependencies for shell app

cogdash replied:
You want to say that somebody who'd like to download the part of this project has to download other pieces?

BlackSoulHub replied:
> You want to say that somebody who'd like to download the part of this project has to download other pieces?

In current realization it's true

cogdash replied:
Okay, it does make sense

BlackSoulHub commented on src/shell/command_handler.rs:
```diff
@@ -0,0 +1,158 @@
+
+pub const OFFICIAL_REPOSITORY: &str = "https://github.com/Reeliks/blaze";
+
+pub struct ShellCommandHandler {
```
I think best way for shell app - move to separate crate (i write about before)

BlackSoulHub commented:
Как мы будем исполнять команды?

Только как функции наподобии этой?

``` ruby
function get_total_cart_price(products: Products): float {
    mut total: float = 0;
    product of products {
        total += product.price;
    };
    total
}
``` 

И как будет организована сама база данных?

Появилась подобная идея:

``` rust 
pub trait DatatabaseDriver {
    fn get_table(&mut self, table_name: &str) -> Result<Option<Table>>, DbError>;
}
``` 

и таблица:

``` rust 
struct Table {
    // fields ...
} 

impl Table {
    // TODO: write pretty API 
}
```

BlackSoulHub commented:
> Добавление абсолютно новых полей, таблиц и лейнов, а также удаление уже существующих проходит без проблем, но есть нюанс с их изменением, вытекающих из того, что структура очень подвижна: названия, типы данных и записанные значения необходимо изменять явно, например так: cash: int -> money: float. Могут быть и моменты, когда автоматического преобразования типов будет недостаточно, и здесь может понадобится ручное написание приведения типов путём пропуска через функции.

Я думаю, что мы должны хранить схему базы данных в где-то в файле и не позволять запускать приложение с невалидной схемой. В конце концов каст данных формата: int  -> float, даже одинаковой битности - некорректная операция.
Для изменения схемы пользователь должен написать какую-нибудь миграцию

> Код будет исполняться согласно структуре базы данных

Интересный момент по поводу кода. То как он выглядит очень похоже на какой-нибудь язык программирования.
Теоретически мы можем написать сервер как интерпретатор этого языка. Но подобная свобода довольно опасна. Теже функции в теории могут вызывать переполнение стека
Нам нужна законченная модель языка

Reeliks commented:
Да, без хранения некоторой информации о последнем состоянии структуры файла данных не обойтись, так как помимо самих данных и их связей необходимо отдельно записывать id-инкременты (если это не uuid), "внешние ключи", ограничения (например, on cascade и restrict) и флаг дефолтного значения. Однако хранение типа отдельным параметром вызывает сомнение, поскольку в планах ввести тип wild, способный хранить любые значения, а также параметр базы данных, при котором миграция является не ещё одним этапом запуска, а процедурой, происходящей при попытке взаимодействия с данными. 

> Теоретически мы можем написать сервер как интерпретатор этого языка.

Ты попал в точку с интерпретатором. Да, будет использоваться новый интерпретируемый язык программирования, схожий по синтаксису на python и java script и имеющий больше возможностей, чем другие языки запросов.

>  Теже функции в теории могут вызывать переполнение стека

Здесь будет применено одно из свойств ACID — изменения будут применены к реальным данным только после успешного завершения кода. Вне самой базы данных (а возможно и в ней тоже) будут доступны коммиты.

Я нахожусь в активных размышлениях по поводу функционала языка и базы данных в целом, мне нужно некоторое время на это. Думаю, нам стоит завести документацию и уже туда складывать устоявшиеся концепты.

Reeliks commented:
Я думаю, такой вариант может стать неплохим решением для миграций:

```r
table machines -> manufacturing.assets.machines {
	name: str?,
	warehouse_id: int = 0
		-> warehouse_number: str!,
	current_factory: str?,
		-get_factory_by_name> current_factories: set[db.manufacturing.factories],
	responsible: db.workers,
	release_year: int[>1925 <2150],
	next_inspection: datetime[utc]?,
	broken: bool = false,
	notes: str
		-notes_to_extra> extra_info: dict[str, str],
};

function get_factory_by_name(factory_name: str): db.manufacturing.factories {
	return db.manufacturing.factories.(name=factory_name)[..0-1]; // [..0-1] must be considered
} // Then the transformation from a link into a set of links happens automatically

function notes_to_extra(notes: str): dict[str, str] {
	mut extra: dict[str, str] = [];
	note of notes.split(r"\n( )*\*") {
		extra.push(note.split(":", max_times=1));
	};
	return extra;
}
```

BlackSoulHub commented:
> при попытке обратиться к полю по прошлому имени произойдёт редирект на новую реализацию

А если в процессе миграций название этого поля было занято?

```
table machines -> manufacturing.assets.machines
```

Это переименование таблицы? Если да, то новое название это что-то типо пространств имён?

Reeliks commented:
При поднятии сервера будет проводиться синтаксический и семантический анализ, которые будут также выявлять дубликацию имён

Да, лейны вполне можно назвать пространствами имён, которые могут использоваться для разделения секций базы данных и управления доступом пользователей

BlackSoulHub commented:
> При поднятии сервера будет проводиться семантический анализ, который будет выявлять дубликацию имён в коде и самом файле с данными

Это выглядит слишком нестрогим. Думаю нужно написать законченнцю версию языка с объяснение синтаксиса и посмотреть на каком-нибудь примере

> Да, лейны вполне можно назвать пространствами имён, которые могут использоваться для разделения секций базы данных и управления доступом пользователей

Звучит интересно, но выгляди довольно сложным. Тут тоже, что и выше. Нужно посмотреть как это будет выглядеть. В конце концов база данных, вроде как, планируется простой. И подобный уровень настройки доступов кажется избыточным
